libibverbs/device.c

```diff
@@ -262,6 +267,8 @@ struct ibv_context *__ibv_open_device(struct ibv_device *device)
 	context->cmd_fd = cmd_fd;
 	pthread_mutex_init(&context->mutex, NULL);
 
+	ibverbs_device_hold(device);
+
```

Comment:
I'm not talking about clearing memory, I am saying ibv_open_device should not return a device we already know to be gone. 

I disagree about the race, it is very relevant to this series and must be addressed here, like you say, by checking the timestamp in proper places. Previously we expressly did not support hotplug while verbs apps are running, now we do, so we can no longer ignore these fringe cases.

You need to think of something to make sysfs safe as well, maybe openat()...